src/pkg/fault/example_fault_test.go

```diff
@@ -0,0 +1,180 @@
+package fault_test
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/alcionai/corso/src/pkg/fault"
+)
+
+// ---------------------------------------------------------------------------
+// mock helpers
+// ---------------------------------------------------------------------------
+
+var (
+	ctrl  any
+	anErr error
+	items = []string{}
+)
+
+type mockController struct {
+	errors any
+}
+
+func connectClient() error      { return nil }
+func dependencyCall() error     { return nil }
+func getIthItem(i string) error { return nil }
+
+// ---------------------------------------------------------------------------
+// examples
+// ---------------------------------------------------------------------------
+
+// ExampleNewErrors highlights assumptions and best practices
+// for generating Errors structs.
+func Example_new() {
+	// Errors should only be generated during the construction of
+	// another controller, such as a new Backup or Restore Operations.
+	// Configurations like failFast are handled during construction.
+	ctrl = mockController{
+		errors: fault.New(false),
+	}
+}
+
+// ExampleErrorsFail describes the assumptions and best practices
+// for setting the Failure error.
+func Example_errors_Fail() {
+	errs := fault.New(false)
+
+	// Fail() should be used to record any error that highlights a
+	// non-recoverable failure in a process.
+	if err := connectClient(); err != nil {
+		errs.Fail(err)
+	}
+
+	// Only the topmost handler of the error should set the Fail() err.
+	// This will normally be the operation controller itself.
+	// IE: Fail() is not Wrap().  In lower levels, errors should get
+	// wrapped and returned like normal, and only handled by errors
+	// at the end.
+	lowLevelCall := func() error {
+		if err := dependencyCall(); err != nil {
+			// wrap here, deeper into the stack
+			return errors.Wrap(err, "dependency")
+		}
+
+		return nil
+	}
+
+	if err := lowLevelCall(); err != nil {
+		// fail here, at the top of the stack
+		errs.Fail(err)
+	}
+}
+
+// ExampleErrorsAdd describes the assumptions and best practices
+// for aggregating iterable or recoverable errors.
+func Example_errors_Add() {
```

Comment:
In "practice" the asymmetry between when `Fail` vs. when `Add` should be called seems a little confusing. The rules I'm coming up with looking at the examples and comments are
* `Fail` at the top-level but return errors like normal
* `Add` in the low-level function and check to see if we should exit

While easily summed up, it seems like it could be confusing as to where to put which. We're now also going to have to check both the returned `error` from a function (i.e. something that will turn into a call to `Fail()` at some point in the stack) _and_ the `fault.errors.Err()` value to see if we should exit a function instead of trying to continue

Reply:
> The rules I'm coming up with looking at the examples and comments are

That's pretty much correct, though I'm sure I can make that more clear.  I'll go do a pass on the language for improvements.

> We're now also going to have to check both the returned error from a function (i.e. something that will turn into a call to Fail() at some point in the stack) and the fault.errors.Err() value to see if we should exit a function instead of trying to continue

No, we shouldn't be doing that.  Only the users at the interface layer (CLI, SDK) need to investigate both Err() and Errs().

Within corso, error handling should only do one of three things:
1. Return an error, if possible (this is most of our error handling).
2. Add() an error, when within some iteration.
3. Fail() an error; only backup/restore operations should do this.

The majority of our error handling should feel entirely standard.  Ie: just create/wrap/return errors, and check the error that was returned.

Reply:
>No, we shouldn't be doing that. Only the users at the interface layer (CLI, SDK) need to investigate both Err() and Errs().

I think we're talking about slightly different things.

Assume you have the following:
```go
func caller() fault.errors {
  errs := fault.NewErrors(true)

  err := callee(&errs)

  // do something with error handling
}
```

How do you check that 1. callee didn't encounter a fatal error and 2. didn't encounter any error while running in fail-fast mode? Wouldn't we by necessity have to check both `fault.errors.Err()` and `err` to determine that? We'd check `Err()` to determine if we failed fast while we'd need to check `err` to see if we encountered a fatal error (e.x. credentials expired)

Reply:
Ahh, I gotcha.

> 1. callee didn't encounter a fatal error

`if errors.Err() != nil`

 > 2. didn't encounter any error while running in fail-fast mode?
 
 `if errors.Err() != nil`
 
 In fail-fast mode, the first recoverable error automatically promotes itself to the non-recoverable error slot.  IE: "this is the error that caused the operation to end".  
 
 In either case, I'd also say the example is a little different from what I'd expect in usage (I have a more complete example in the examples test now, but to try and shorten it here:)
 
 ```go
 func caller() fault.errors {
  errs := fault.NewErrors(true)

  if err := callee(&errs); err != nil {
    return errs.Fail(err)
  }

  recoverableCallee(&errs)
  
  return errs
}
```

And I realize that we'd need to check both the callee's `if err != nil && errs.Err() != nil`.  So a QoL helper could ease the boilerplate a lot.  I'll think on approaches there.